proto/asn1-pdu/asn1_universal_types_to_der.cc

```diff
@@ -67,6 +67,39 @@ void Encode(const BitString& bit_string, std::vector<uint8_t>& der) {
   }
 }
 
+void Encode(const ObjectIdentifier& object_identifier,
+            std::vector<uint8_t>& der) {
+  // Save the current size in |tag_len_pos| to place tag and length
+  // after the value is encoded.
+  const size_t tag_len_pos = der.size();
+
+  uint8_t root = object_identifier.root();
+  uint8_t small_identifier = object_identifier.small_identifier();
+  auto subidentifier = object_identifier.subidentifier();
+
+  // (X.690 (2015) 8.19.4): Only 39 subsequent values from nodes reached by X =
+  // 0 and X = 1. Therefore, use |small_identifier| for |root| 0 or 1, and when
+  // |root| is 2, use first integer in |subidentifier| to obtain
+  // potentially higher values.
+  size_t identifier = (root == 2 && !subidentifier.empty())
+                          ? (root * 40) + subidentifier[0]
+                          : (root * 40) + small_identifier;
```

Comment:
```suggestion
  size_t identifier = (root * 40) + small_identifier;
  if (root == 2 && !subidentifier.empty()) {
    identifier += subidentifier.back();
    subidentifier.pop_back();
  }
```

Maybe this? I mention it only because you end up duplicating the use of `subidentifier[0]` for the child subidentifiers, meaning you'll always end up with the 2nd & 3rd identifier repeated. The above changes to grab the 'last' identifier (since pop_back() is basically free and just a size truncation without actually freeing the memory), so that we end up with distinct values. It still consumes small_identifier, to make it a little easier to read.

That said, looking at http://oid-info.com/get/2 the valid set of OIDs within the root arc 2 are low. But I think you're spot on for making sure this is robust enough to handle long-form first-octets as part of fuzzing, so I *think* the above approach might be a little bit cleaner/more random.